service/src/event/ordering_task.rs

```diff
@@ -175,599 +125,292 @@ impl OrderingTask {
-        let mut stream = Self::new();
-        for item in iter {
-            stream.add_event(item);
-        }
-        stream
-    }
+    cid_map: HashMap<EventCid, StreamCommit>,
 }
 
 impl StreamEvents {
-    pub fn new() -> Self {
-        Self::default()
+    fn new<I>(cid: StreamCid, events: I) -> Self
+    where
+        I: IntoIterator<Item = StreamCommit>,
+    {
+        let mut new = Self::new_empty(cid);
+
+        for event in events {
+            new.add_event(event);
+        }
+        new
     }
 
-    /// returns Some(Stream Init CID) if this is a new event, else None.
-    pub fn add_event(&mut self, event: DeliverableEvent) -> Option<InitCid> {
-        let res = if self.prev_map.insert(event.meta.prev, event.cid).is_none() {
-            Some(event.meta.init_cid)
-        } else {
-            None
-        };
-        self.cid_map.insert(event.cid, event);
-        res
+    fn new_empty(cid: StreamCid) -> Self {
+        Self {
+            _cid: cid,
+            prev_map: HashMap::default(),
+            cid_map: HashMap::default(),
+        }
     }
 
-    pub fn is_empty(&self) -> bool {
-        // these should always match
-        self.prev_map.is_empty() && self.cid_map.is_empty()
+    /// returns true if this is a new event.
+    fn add_event(&mut self, event: StreamCommit) -> bool {
+        if let Some(prev) = event.prev {
+            self.prev_map.insert(prev, event.cid);
+        }
+        self.cid_map.insert(event.cid, event).is_none()
     }
 
-    fn remove_by_event_cid(&mut self, cid: &Cid) -> Option<DeliverableEvent> {
-        if let Some(cid) = self.cid_map.remove(cid) {
-            self.prev_map.remove(&cid.meta.prev);
-            Some(cid)
+    fn remove_by_event_cid(&mut self, cid: &Cid) -> Option<StreamCommit> {
+        if let Some(ev) = self.cid_map.remove(cid) {
+            if let Some(prev) = ev.prev {
+                self.prev_map.remove(&prev);
+            }
+            Some(ev)
         } else {
             None
         }
     }
 
-    fn remove_by_prev_cid(&mut self, cid: &Cid) -> Option<EventCid> {
-        if let Some(cid) = self.prev_map.remove(cid) {
+    fn remove_by_prev_cid(&mut self, prev: &Cid) -> Option<EventCid> {
+        if let Some(cid) = self.prev_map.remove(prev) {
             self.cid_map.remove(&cid);
             Some(cid)
         } else {
             None
         }
     }
+
+    fn delivered_events(&self) -> impl Iterator<Item = &EventCid> {
+        self.cid_map
+            .iter()
+            .filter_map(|(cid, event)| if event.delivered { Some(cid) } else { None })
+    }
+
+    fn order_events(&mut self, start_with: EventCid) -> VecDeque<Cid> {
+        let mut deliverable = VecDeque::with_capacity(self.prev_map.len());
+
+        deliverable.push_back(start_with);
+        self.remove_by_event_cid(&start_with);
+        let mut tip = start_with;
+        // technically, could be in deliverable set as well if the stream is forking?
```

Comment:
we should definitely make sure we have tests that cover the case of stream forks and multiple histories each with gaps that get filled in at different times.